File: packages/nitro-renderer/src/avatar/AvatarStructure.ts
import { AvatarBodyPartType, AvatarDirectionAngle, AvatarFigurePartType, AvatarGenderType, AvatarGeometryType, AvatarScaleType, AvatarSetType, IActionDefinition, IActiveActionData, IAnimation, IAnimationLayerData, IAssetAnimation, IAssetAvatarActionData, IAssetAvatarAnimation, IAssetAvatarBodyPartItem, IAssetAvatarGeometryConfig, IAssetAvatarPartSetItem, IAssetAvatarPartSets, IAvatarCanvas, IAvatarFigureContainer, IAvatarImage, IAvatarStructure, IFigureData, IFigurePartSet, IPartColor, IStructureData } from '@nitrodevco/nitro-api';
import { Point } from 'pixi.js';

import { ActionDefinition, AvatarActionManager } from './actions';
import { AnimationManager } from './animation';
import { AvatarImagePartContainer } from './AvatarImagePartContainer';
import { AvatarModelGeometry } from './geometry';
import { AnimationAction, AvatarAnimationData, FigureSetData, PartSetsData } from './structure';

export class AvatarStructure implements IAvatarStructure {
    private _geometry: AvatarModelGeometry | undefined = undefined;
    private _figureData: FigureSetData = new FigureSetData();
    private _partSetsData: PartSetsData = new PartSetsData();
    private _animationData: AvatarAnimationData = new AvatarAnimationData();
    private _animationManager: AnimationManager = new AnimationManager();
    private _actionManager: AvatarActionManager = new AvatarActionManager();
    private _mandatorySetTypeIds: Map<AvatarGenderType, Record<number, AvatarFigurePartType[]>> = new Map();
    private _defaultAction: ActionDefinition | undefined = undefined;
    private _defaultLayAction: ActionDefinition | undefined = undefined;

    /** Flash `init()`: called once the figure data has been downloaded, so the mandatory-set cache is rebuilt from it. */
    public init(): void {
        this._mandatorySetTypeIds = new Map();
    }

    public initGeometry(data: IAssetAvatarGeometryConfig): void {
        this._geometry = new AvatarModelGeometry(data);
    }

    /** `initActions` / `updateActions`: (re)loads the action definitions and picks up the default and default-lay actions. */
    public updateActions(data: IAssetAvatarActionData): void {
        this._actionManager.updateActions(data);

        this._defaultAction = this._actionManager.getDefaultAction();
        this._defaultLayAction = this._actionManager.getDefaultLayAction();
    }

    public initPartSets(data: IAssetAvatarPartSets): boolean {
        if (this._partSetsData && this._partSetsData.parse(data)) {
            const right = this._partSetsData.getPartDefinition('ri');
            const left = this._partSetsData.getPartDefinition('li');

            if (right) right.appendToFigure = true;
            if (left) left.appendToFigure = true;

            return true;
        }

        return false;
    }

    public initAnimation(data: IAssetAvatarAnimation[]): boolean {
        return this._animationData.parse(data);
    }

    public initFigureData(data: IFigureData): boolean {
        return this._figureData.parse(data);
    }

    public injectFigureData(data: IFigureData): void {
        this._figureData.injectJSON(data);
    }

    public registerAnimations(data: IAssetAnimation[]): void {
        this._animationManager.registerAnimations(this, data);
    }

    public registerAnimation(data: IAssetAnimation): void {
        this._animationManager.registerAnimations(this, [ data ]);
    }

    public getPartColor(container: IAvatarFigureContainer, partType: AvatarFigurePartType, layerId: number = 0): IPartColor | undefined {
        const colorIds = container.getPartColorIds(partType);

        if (!colorIds || colorIds.length < layerId) return undefined;

        const figureDataSetType = this._figureData.getSetType(partType);

        if (!figureDataSetType) return undefined;

        return this._figureData.getPalette(figureDataSetType.paletteId)?.getColor(colorIds[layerId]);
    }

    public getBodyPartData(animation: string, frameCount: number, spriteId: string): IAnimationLayerData | undefined {
        return this._animationManager.getLayerData(animation, frameCount, spriteId);
    }

    public getAnimation(name: string): IAnimation | undefined {
        return this._animationManager.getAnimation(name);
    }

    public getActionDefinition(id: string): ActionDefinition | undefined {
        return this._actionManager.getActionDefinition(id);
    }

    public getDefaultActionDefinition(): ActionDefinition | undefined {
        return this._defaultAction;
    }

    public getDefaultLayActionDefinition(): ActionDefinition | undefined {
        return this._defaultLayAction;
    }

    public getActionDefinitionWithState(state: string): ActionDefinition | undefined {
        return this._actionManager.getActionDefinitionWithState(state);
    }

    public isMainAvatarSet(id: AvatarSetType): boolean {
        return this._geometry?.isMainAvatarSet(id) ?? false;
    }

    public sortActions(actions: IActiveActionData[]): IActiveActionData[] {
        return this._actionManager.sortActions(actions);
    }

    public maxFrames(actions: IActiveActionData[]): number {
        let count = 0;

        for (const action of actions) {
            if (!action.definition) continue;

            count = Math.max(count, this._animationData.getFrameCount(action.definition));
        }

        return count;
    }

    public getMandatorySetTypeIds(gender: AvatarGenderType, _arg_2: number): AvatarFigurePartType[] {
        let get = this._mandatorySetTypeIds.get(gender);

        if (!get) {
            get = {};

            this._mandatorySetTypeIds.set(gender, get);
        }

        let value = get[_arg_2];

        if (value === undefined) {
            value = this._figureData.getMandatorySetTypeIds(gender, _arg_2);

            get[_arg_2] = value;
        }

        return value;
    }

    public getDefaultPartSet(partType: AvatarFigurePartType, gender: AvatarGenderType): IFigurePartSet | undefined {
        return this._figureData.getDefaultPartSet(partType, gender);
    }

    public getCanvasOffsets(actions: IActiveActionData[], size: AvatarScaleType, direction: number): number[] {
        return this._actionManager.getCanvasOffsets(actions, size, direction);
    }

    public getCanvas(scaleType: AvatarScaleType, geometryType: AvatarGeometryType): IAvatarCanvas | undefined {
        return this._geometry?.getCanvas(scaleType, geometryType);
    }

    public removeDynamicItems(avatar: IAvatarImage): void {
        this._geometry?.removeDynamicItems(avatar);
    }

    /*
     * The body parts an action touches: for an animation, the parts its frames move, the
     * parts of anything it removes, and the parts its adds attach to (registering those adds
     * as dynamic geometry items); otherwise the parts holding the action's active part set.
     */
    public getActiveBodyPartIds(k: IActiveActionData, avatar: IAvatarImage): AvatarBodyPartType[] {
        if (!k?.definition || !this._geometry) return [];

        const partIds: AvatarBodyPartType[] = [];
        const geometryType = k.definition.geometryType;

        if (k.definition.isAnimation) {
            const animation = this._animationManager.getAnimation(((k.definition.state + '.') + k.actionParameter));

            if (animation) {
                const animatedPartIds = animation.getAnimatedBodyPartIds(0, k.overridingAction);

                for (const remove of animation.removeData) {
                    const bodyPart = this._geometry.getBodyPartOfItem(geometryType, remove, avatar);

                    if (bodyPart && (partIds.indexOf(bodyPart.id) === -1)) partIds.push(bodyPart.id);
                }

                if (animation.hasAddData()) {
                    for (const add of animation.addData) {
                        if (!add.align) continue;

                        const bodyPart = this._geometry.getBodyPart(geometryType, add.align);

                        if (!bodyPart) continue;

                        const partItem = {
                            id: add.id,
                            x: 0,
                            y: 0,
                            z: 0,
                            radius: 0.01,
                            nx: 0,
                            ny: 0,
                            nz: -1,
                            double: true,
                        } as IAssetAvatarBodyPartItem;

                        bodyPart.addPart(partItem, avatar);

                        const partDefinition = this._partSetsData.addPartDefinition({ setType: add.id } as IAssetAvatarPartSetItem);

                        partDefinition.appendToFigure = true;

                        if (add.base === '') partDefinition.staticId = 1;

                        if (partIds.indexOf(bodyPart.id) === -1) partIds.push(bodyPart.id);
                    }
                }

                for (const partId of animatedPartIds) {
                    const bodyPart = this._geometry.getBodyPart(geometryType, partId);

                    if (bodyPart && (partIds.indexOf(bodyPart.id) === -1)) partIds.push(bodyPart.id);
                }
            }
        } else {
            const animatedPartIds = this._partSetsData.getActiveParts(k.definition);

            for (const partId of animatedPartIds) {
                const bodyPart = this._geometry.getBodyPartOfItem(geometryType, partId, avatar);

                if (bodyPart && (partIds.indexOf(bodyPart.id) === -1)) partIds.push(bodyPart.id);
            }
        }

        return partIds;
    }

    public getBodyPartsUnordered(setType: AvatarSetType): AvatarBodyPartType[] {
        return this._geometry?.getBodyPartIdsInAvatarSet(setType) ?? [];
    }

    public getBodyParts(setType: AvatarSetType, geometryType: AvatarGeometryType, direction: number): AvatarBodyPartType[] {
        return this._geometry?.getBodyPartsAtAngle(setType, AvatarDirectionAngle.DIRECTION_TO_ANGLE[direction], geometryType) ?? [];
    }

    public getFrameBodyPartOffset(k: IActiveActionData, direction: number, frame: number, bodyPartId: AvatarBodyPartType): Point {
        if (k.definition) return this._animationData.getAction(k.definition)?.getFrameBodyPartOffset(direction, frame, bodyPartId) ?? AnimationAction.DEFAULT_OFFSET;

        return AnimationAction.DEFAULT_OFFSET;
    }

    public getParts(setType: AvatarBodyPartType, container: IAvatarFigureContainer, activeAction: IActiveActionData, geometryType: AvatarGeometryType, direction: number, removes: string[], avatar: IAvatarImage, layerItems: Map<AvatarFigurePartType, string> = new Map()): AvatarImagePartContainer[] {
        if (!activeAction?.definition || !this._geometry) return [];

        const activeParts = this._partSetsData.getActiveParts(activeAction.definition);
        const animationAction = this._animationData.getAction(activeAction.definition);

        let animation: IAnimation | undefined = undefined;
        let emptyFrames: number[] = [ 0 ];

        if (activeAction.definition.isAnimation) {
            animation = this._animationManager.getAnimation(`${activeAction.definition.state}.${activeAction.actionParameter}`);

            if (animation) {
                emptyFrames = this.getPopulatedArray(animation.frameCount(activeAction.overridingAction));

                for (const bodyPartId of animation.getAnimatedBodyPartIds(0, activeAction.overridingAction)) {
                    if (bodyPartId !== setType) continue;

                    const bodyPart = this._geometry.getBodyPart(geometryType, bodyPartId);

                    if (bodyPart) for (const item of bodyPart.getDynamicParts(avatar)) activeParts.push(item.id);
                }
            }
        }

        const requiredPartTypes = this._geometry.getParts(geometryType, setType, direction, activeParts, avatar);
        const baseContainers: AvatarImagePartContainer[] = [];

        // a part outside the action's active set is drawn in its rest pose: the default action, or its lying-down twin
        const restAction: IActionDefinition = ((activeAction.definition.geometryType === AvatarGeometryType.Horizontal) ? this._defaultLayAction : this._defaultAction) ?? activeAction.definition;

        for (const figurePartType of container.getPartTypeIds()) {
            if (layerItems && layerItems.get(figurePartType) !== undefined) continue;

            const partSetType = this._figureData.getSetType(figurePartType);
            const partSetId = container.getPartSetId(figurePartType);
            const partColorIds = container.getPartColorIds(figurePartType);

            if (!partSetType) continue;

            const palette = this._figureData.getPalette(partSetType.paletteId);
            const partSet = partSetType.getPartSet(partSetId);

            if (!palette || !partSet) continue;

            removes = removes.concat(partSet.hiddenLayers);

            for (const part of partSet.parts) {
                if (requiredPartTypes.indexOf(part.type) === -1) continue;

                const animationFrames = animationAction?.getPart(part.type)?.frames ?? emptyFrames;
                const actionDefinition = (activeParts.indexOf(part.type) === -1) ? restAction : activeAction.definition;
                const partDefinition = this._partSetsData.getPartDefinition(part.type);

                let flippedPartType = !partDefinition ? part.type : partDefinition.flippedSetType;

                if (!flippedPartType) flippedPartType = part.type;

                let partColor: IPartColor | undefined = undefined;

                if (partColorIds && (partColorIds.length > (part.colorLayerIndex - 1))) partColor = palette.getColor(partColorIds[(part.colorLayerIndex - 1)]);

                baseContainers.push(new AvatarImagePartContainer(setType, part.type, part.id.toString(), partColor, animationFrames, actionDefinition, (part.colorLayerIndex > 0), part.paletteMap, flippedPartType));
            }
        }

        const partContainers: AvatarImagePartContainer[] = [];

        for (const partType of requiredPartTypes) {
            let partColor: IPartColor | undefined = undefined;
            let hasFigurePart = false;

            const layerItem = layerItems.get(partType);
            const hasLayerItem = (layerItem !== undefined);

            for (const partContainer of baseContainers) {
                if (partContainer.partType !== partType) continue;

                if (hasLayerItem) {
                    partColor = partContainer.color;
                } else {
                    hasFigurePart = true;

                    if (removes.indexOf(partType) === -1) partContainers.push(partContainer);
                }
            }

            if (hasFigurePart) continue;

            if (hasLayerItem) {
                // an animation item replaces the figure's part with the asset named by its base
                partContainers.push(new AvatarImagePartContainer(setType, partType, layerItem, partColor, animationAction?.getPart(partType)?.frames ?? emptyFrames, activeAction.definition, !!partColor, -1, partType, false, 1));
            } else if (activeParts.indexOf(partType) > -1) {
                const bodyPart = this._geometry.getBodyPartOfItem(geometryType, partType, avatar);

                if (bodyPart && (setType !== bodyPart.id)) continue;

                const partDefinition = this._partSetsData.getPartDefinition(partType);

                if (!partDefinition || !partDefinition.appendToFigure) continue;

                let isBlended = false;
                let blend = 1;
                let partId = '1';

                if (activeAction.actionParameter !== '') partId = activeAction.actionParameter;

                if (partDefinition.hasStaticId()) partId = partDefinition.staticId.toString();

                if (animation) {
                    const addData = animation.getAddData(partType);

                    if (addData) {
                        isBlended = addData.isBlended;
                        blend = addData.blend;
                    }
                }

                partContainers.push(new AvatarImagePartContainer(setType, partType, partId, undefined, animationAction?.getPart(partType)?.frames ?? emptyFrames, activeAction.definition, false, -1, partType, isBlended, blend));
            }
        }

        return partContainers;
    }

    private getPopulatedArray(size: number): number[] {
        const arr: number[] = [];

        let index = 0;

        while (index < size) {
            arr.push(index);

            index++;
        }

        return arr;
    }

    /** The CarryItem parameter ids: every item id the client can hold. */
    public getItemIds(): string[] {
        const definition = this._actionManager.getActionDefinition('CarryItem');

        if (!definition) return [];

        return definition.params.keys().toArray();
    }

    public get actionManager(): AvatarActionManager {
        return this._actionManager;
    }

    public get figureData(): IStructureData {
        return this._figureData;
    }

    public get partData(): PartSetsData {
        return this._partSetsData;
    }

    public get animationManager(): AnimationManager {
        return this._animationManager;
    }
}

File: packages/nitro-swift/Sources/NitroAvatar/AvatarStructure.swift
import Foundation

/// Swift port of `AvatarStructure` (packages/nitro-renderer/src/avatar/AvatarStructure.ts) -
/// the figure+action -> visible-layers resolver, and the top-level owner of the figure-data
/// catalog, part-sets table, geometry config, action catalog and keyframe animation data.
///
/// Reduced scope vs. the TS source (see the avatar section of the package README): no FX/gesture
/// "add data" layer injection and no downloaded-effects animation system (`AnimationManager`/
/// `Animation`, the "Dance"/"Effect" states) - `getActiveBodyPartIds`/`getParts` below only take the
/// non-effects path, which is everything the 10 bundled `HabboAvatarAnimations.json` entries
/// (Default/Sit/Lay/Move/Wave/Talk/Sign/Respect/Blow/Laugh) actually need, since none of their
/// `ActionDefinition`s set `isAnimation` (that flag is reserved for effects-bundle-backed states).
public final class AvatarStructure {
    private var geometry: AvatarModelGeometry?
    public let figureData = FigureSetData()
    public let partData = PartSetsData()
    public let animationData = AvatarAnimationData()
    public let actionManager = AvatarActionManager()

    public init() {}

    @discardableResult
    public func initAnimation(_ data: [AvatarAnimationConfig]) -> Bool { animationData.parse(data) }

    /// Additive, like `injectFigureData` - see `AvatarActionManager.updateActions`'s doc comment for
    /// why a host app needs this to unlock anything beyond the bundled "Stand" action.
    public func registerActionData(_ data: AvatarActionDataConfig) { actionManager.updateActions(data) }

    public func getActionDefinition(_ id: String) -> ActionDefinition? { actionManager.getActionDefinition(id) }

    public func getActionDefinitionWithState(_ state: String) -> ActionDefinition? {
        actionManager.getActionDefinitionWithState(state)
    }

    public func getDefaultAction() -> ActionDefinition? { actionManager.getDefaultAction() }

    /// See `AvatarActionManager.sortActions`'s doc comment - despite the name, this does not sort by
    /// precedence (a faithfully-replicated bug in the original).
    public func sortActions(_ actions: [ActiveActionData]) -> [ActiveActionData] { actionManager.sortActions(actions) }

    public func maxFrames(_ actions: [ActiveActionData]) -> Int {
        var count = 0

        for action in actions {
            guard let definition = action.definition else { continue }

            count = max(count, animationData.getFrameCount(definition))
        }

        return count
    }

    public func getCanvasOffsets(_ actions: [ActiveActionData], scale: AvatarScaleType, direction: Int) -> (Double, Double, Double)? {
        actionManager.getCanvasOffsets(actions, size: scale, direction: direction)
    }

    /// Which body parts a given active action claims - the non-effects path of `getActiveBodyPartIds`
    /// (see the class doc comment). Effects/dance actions (`definition.isAnimation`) claim nothing
    /// here, matching an avatar with no effects bundle registered for them.
    public func getActiveBodyPartIds(_ action: ActiveActionData) -> [AvatarBodyPartType] {
        guard let definition = action.definition, let geometry, !definition.isAnimation else { return [] }

        var partIds: [AvatarBodyPartType] = []

        for partId in partData.getActiveParts(definition.activePartSet) {
            guard let bodyPart = geometry.getBodyPartOfItem(geometryType: definition.geometryType, itemId: partId) else { continue }

            if !partIds.contains(bodyPart) { partIds.append(bodyPart) }
        }

        return partIds
    }

    /// The per-frame pixel nudge for one body part under one active action - `AnimationAction`'s
    /// own keyframe-offset table (see its doc comment), or `(0, 0)` for actions with no such table.
    /// Of the 10 bundled `HabboAvatarAnimations.json` entries, only "Laugh" declares an `offsets`
    /// block (an 8-direction head bob) - every other bundled animation relies solely on per-part
    /// keyframe *frame numbers* (`AnimationActionPart`), not pixel offsets.
    public func getFrameBodyPartOffset(_ action: ActiveActionData, direction: Int, frame: Int, bodyPartId: AvatarBodyPartType) -> (dx: Double, dy: Double) {
        guard let definition = action.definition else { return AnimationAction.defaultOffset }

        return animationData.getAction(definition)?.getFrameBodyPartOffset(direction: direction, frameCount: frame, bodyPartId: bodyPartId)
            ?? AnimationAction.defaultOffset
    }

    public func initGeometry(_ config: AvatarGeometryConfig) {
        geometry = AvatarModelGeometry(config: config)
    }

    /// Mirrors `AvatarStructure.initPartSets`: after parsing, the mirrored hand-item slots ("ri"/"li")
    /// are marked `appendToFigure` so `getParts` can synthesize a held-item layer even when the
    /// figure string itself has no "ri"/"li" segment.
    @discardableResult
    public func initPartSets(_ config: AvatarPartSetsConfig) -> Bool {
        guard partData.parse(config) else { return false }

        partData.getPartDefinition(.rightHandItem)?.appendToFigure = true
        partData.getPartDefinition(.leftHandItem)?.appendToFigure = true

        return true
    }

    @discardableResult
    public func initFigureData(_ data: FigureData) -> Bool { figureData.parse(data) }

    public func injectFigureData(_ data: FigureData) { figureData.appendJSON(data) }

    public func getPartColor(_ container: AvatarFigureContainer, partType: AvatarFigurePartType, layerId: Int = 0) -> PartColor? {
        let colorIds = container.getPartColorIds(partType)

        guard colorIds.count >= layerId, let setType = figureData.getSetType(partType) else { return nil }

        return figureData.getPalette(setType.paletteId)?.getColor(colorIds[layerId])
    }

    public func isMainAvatarSet(_ setType: AvatarSetType) -> Bool { geometry?.isMainAvatarSet(setType) ?? false }

    public func getMandatorySetTypeIds(gender: AvatarGenderType, count: Int) -> [AvatarFigurePartType] {
        figureData.getMandatorySetTypeIds(gender: gender, count: count)
    }

    public func getDefaultPartSet(_ partType: AvatarFigurePartType, gender: AvatarGenderType) -> FigurePartSet? {
        figureData.getDefaultPartSet(partType, gender: gender)
    }

    public func getCanvas(scale: AvatarScaleType, geometryType: AvatarGeometryType) -> AvatarCanvasInfo? {
        geometry?.getCanvas(scale: scale, geometryType: geometryType)
    }

    /// Body parts belonging to `setType`, nearest-camera-first at the given 0-7 facing direction.
    public func getBodyParts(setType: AvatarSetType, geometryType: AvatarGeometryType, direction: Int) -> [AvatarBodyPartType] {
        geometry?.getBodyPartsAtAngle(setType: setType, direction: direction, geometryType: geometryType) ?? []
    }

    public func getBodyPartsUnordered(_ setType: AvatarSetType) -> [AvatarBodyPartType] {
        geometry?.getBodyPartIds(inAvatarSet: setType) ?? []
    }

    /// Swift port of `AvatarStructure.getParts` - see the type doc comment for what's out of scope
    /// (the `layerItems`/effects-animation parameters from the original signature are dropped
    /// entirely, not just defaulted, since nothing in this port ever populates them).
    public func getParts(
        setType: AvatarBodyPartType,
        container: AvatarFigureContainer,
        activeAction: ActiveActionData,
        geometryType: AvatarGeometryType,
        direction: Int,
        removes initialRemoves: [String] = []
    ) -> [AvatarImagePartContainer] {
        guard let geometry, let definition = activeAction.definition else { return [] }

        let activeParts = partData.getActiveParts(definition.activePartSet)
        let emptyFrames: [AvatarFrameEntry] = [.index(0)]
        let requiredPartTypes = geometry.getParts(geometryType: geometryType, bodyPartId: setType, direction: direction)
        let animationAction = animationData.getAction(definition)

        var removes = initialRemoves
        var baseContainers: [AvatarImagePartContainer] = []

        for figurePartType in container.getPartTypeIds() {
            guard let partSetType = figureData.getSetType(figurePartType) else { continue }

            let partSetId = container.getPartSetId(figurePartType)
            let partColorIds = container.getPartColorIds(figurePartType)

            guard
                let palette = figureData.getPalette(partSetType.paletteId),
                let partSet = partSetType.getPartSet(partSetId)
            else { continue }

            removes.append(contentsOf: partSet.hiddenLayers)

            for part in partSet.parts {
                guard requiredPartTypes.contains(part.type) else { continue }

                // `AvatarStructure._defaultAction` is declared in the TS source but never assigned
                // anywhere, so its "fall back to the default action for parts outside activeParts"
                // branch is always a no-op in practice - `activeAction` is used unconditionally here.
                let partDefinition = partData.getPartDefinition(part.type)
                let flippedPartType = partDefinition?.flippedSetType ?? part.type

                var partColor: PartColor?
                let colorSlot = part.colorLayerIndex - 1

                if colorSlot >= 0, partColorIds.count > colorSlot { partColor = palette.getColor(partColorIds[colorSlot]) }

                let animationFrames = animationAction?.getPart(part.type).map { $0.frames.map(AvatarFrameEntry.keyframe) } ?? emptyFrames

                baseContainers.append(AvatarImagePartContainer(
                    bodyPartId: setType.rawValue, partType: part.type, partId: part.id, color: partColor,
                    frames: animationFrames, action: definition, isColorable: part.colorLayerIndex > 0,
                    paletteMapId: 0, flippedPartType: flippedPartType
                ))
            }
        }

        var partContainers: [AvatarImagePartContainer] = []

        for partType in requiredPartTypes {
            var matchedFromFigure = false

            for baseContainer in baseContainers where baseContainer.partType == partType {
                matchedFromFigure = true

                if !removes.contains(partType.rawValue) { partContainers.append(baseContainer) }
            }

            guard !matchedFromFigure, activeParts.contains(partType) else { continue }

            guard let partDefinition = partData.getPartDefinition(partType), partDefinition.appendToFigure else { continue }

            // `activeAction.actionParameter` (default 1) picks e.g. which carried-item variant to
            // show; a static id declared on the part definition itself always wins.
            let partId = partDefinition.hasStaticId() ? partDefinition.staticId : activeAction.actionParameter
            let animationFrames = animationAction?.getPart(partType).map { $0.frames.map(AvatarFrameEntry.keyframe) } ?? emptyFrames

            partContainers.append(AvatarImagePartContainer(
                bodyPartId: setType.rawValue, partType: partType, partId: partId, color: nil,
                frames: animationFrames, action: definition, isColorable: false, paletteMapId: -1,
                flippedPartType: partType, isBlendable: false
            ))
        }

        return partContainers
    }
}
